### ISSUE TITLE
MultiprocessingIterator Deadlock

### ISSUE THREAD

**Issue by thuyen** [chainer/chainer#2909]
I am new to chainer and trying to do a custom loop with MultiprocessingIterator. 

Reproduce on Ubuntu `16.04` and Chainer version `2.0`
````
import chainer
import numpy as np

dataset = np.ones((10000, 1000))
loader = chainer.iterators.MultiprocessIterator(
                dataset, 32, n_processes=6,
                shuffle=True, repeat=True)

while loader.epoch < 1:
    x = loader.next()
print('test')
loader.finalize() # hang at worker.join() (randomly, very frequently)
````

**Comment by niboshi** [comment]
Thank you for reporting!

**Comment by delta2323** [comment]
@niboshi Could you try CuPy 1.0.1 as well to see if this problem is related to #3075?

**Comment by niboshi** [comment]
@delta2323 I tested in these combinations. This problem is reproduced in both conditions.

- Chainer v2.0.2 (`v2`) - CuPy v1.0.2 (`v1`)
- Chainer v3.0.0b1 (`master`) - CuPy v2.0.0b1 (`master`)

**Comment by delta2323** [comment]
Thank you. The deadlock of #3075 occurred with the combination of v2.0.1 and v2.0.0b1 though. Could you try that if you can afford to do that?

**Comment by delta2323** [comment]
Thank you. I will ask them to try master branches for both modules.

**Comment by okuta** [comment]
@thuyen I tested your code with Chainer v3.0.0rc1. The deadlock does not occur.
Maybe, This bug was fixed by #3076.
Cloud you check on your environment?

**Comment by niboshi** [comment]
@okuta 
Ah yes, this must have been solved, because #3076 includes a test case to check for this issue.
(https://github.com/chainer/chainer/pull/3076/files#diff-f4b0901816a48909da3e084290dbb1fcR394)

**Comment by okuta** [comment]
OK, I close this Issue.
